sw/device/silicon_creator/lib/cert/asn1.c

```diff
@@ -0,0 +1,172 @@
+  if (tag->state == NULL)
+    return kErrorAsn1Internal;
+  // Compute actually used length.
+  size_t length = tag->state->offset - tag->len_offset - tag->len_size;
+  // Make sure the length is not too large.
+  if (length > 0xffff)
+    return kErrorAsn1Internal;
+  // If the length requires two bytes, then there is no need to move the data.
+  if (length > 0x7f) {
+    // Sanity check: asn1_start_tag should have outputted three bytes.
+    if (tag->len_size != 3)
+      return kErrorAsn1Internal;
+    // Fixup length.
+    tag->state->buffer[tag->len_offset + 0] = 0x82;
+    tag->state->buffer[tag->len_offset + 1] = (uint8_t)(length >> 8);
+    tag->state->buffer[tag->len_offset + 2] = (uint8_t)(length & 0xff);
+  } else {
+    // We only need one byte to hold the length but we allocated three:
+    // we need to shift all the tag data by two bytes. A forward copy
+    // works for that purpose.
+    // TODO: use memmove when we have an implementation.
+    for (size_t i = 0; i < length; i++) {
+      tag->state->buffer[tag->len_offset + 1 + i] =
+          tag->state->buffer[tag->len_offset + tag->len_size + i];
+    }
+    // Fixup length.
+    tag->state->buffer[tag->len_offset] = (uint8_t)length;
+    // Fixed up state offset.
+    tag->state->offset -= tag->len_size - 1;
+  }
+  // Hardening: clear out the tag structure to prevent accident reuse.
+  tag->state = NULL;
+  tag->len_offset = 0;
+  tag->len_size = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_bool(asn1_state_t *state, bool value) {
+  // Avoid using the tag functions since everything is constant.
+  RETURN_IF_ERROR(asn1_push_byte(state, ASN1_TAG_BOOLEAN));
+  RETURN_IF_ERROR(asn1_push_byte(state, 1));
+  return asn1_push_byte(state, value ? 0xff : 0);
+}
+
+rom_error_t asn1_push_int32(asn1_state_t *state, uint8_t tag, int32_t value) {
+  uint32_t u_value = (uint32_t)value;
+  uint8_t bigint[4] = {
+      u_value >> 24,
+      (u_value >> 16) & 0xff,
+      (u_value >> 8) & 0xff,
+      u_value & 0xff,
+  };
+  return asn1_push_integer(state, tag, true, bigint, sizeof(bigint));
+}
+
+rom_error_t asn1_push_uint32(asn1_state_t *state, uint8_t tag, uint32_t value) {
+  uint8_t bigint[4] = {
+      value >> 24,
+      (value >> 16) & 0xff,
+      (value >> 8) & 0xff,
+      value & 0xff,
+  };
+  return asn1_push_integer(state, tag, false, bigint, sizeof(bigint));
+}
+
+rom_error_t asn1_push_integer(asn1_state_t *state, uint8_t tag, bool is_signed,
+                              const uint8_t *bytes_be, size_t size) {
+  if (size == 0)
+    return kErrorAsn1InvalidArgument;
+  asn1_tag_t tag_st;
+  RETURN_IF_ERROR(asn1_start_tag(state, &tag_st, tag, size));
+  // Compute smallest possible encoding: ASN1 forbids that the first 9 bits (ie
+  // first octet) and MSB of the second octet are either all ones or all zeroes.
+  size_t offset = 0;
+  // First get rid of unecessary zeroes.
+  while (size >= 2 && bytes_be[offset] == 0 &&
```

Comment:
I personally think the logic might be easier to follow if  the pointers are adjusted themselves instead of using offsets, but I'll leave the decision to you.

Reply:
Ok, I have changed it to pointers